test/cadet_web/controllers/answer_controller_test.exs

```diff
@@ -1,10 +1,185 @@
+      assert response(programming_conn, 400) == "Missing or invalid parameter(s)"
+      assert is_nil(get_answer_value(programming_question, assessment, user))
+    end
+  end
+
+  @tag authenticate: :student
+  test "invalid params missing question is unsuccessful", %{
+    conn: conn,
+    assessment: assessment,
+    mcq_question: mcq_question
+  } do
+    user = conn.assigns.current_user
+    {:ok, _} = Repo.delete(mcq_question)
+
+    conn = post(conn, build_url(mcq_question.id), %{answer: 5})
+    assert response(conn, 400) == "Question not found"
+    assert is_nil(get_answer_value(mcq_question, assessment, user))
+  end
+
+  @tag authenticate: :student
+  test "invalid params not open submission is unsuccessful", %{conn: conn} do
+    user = conn.assigns.current_user
+
+    before_open_at_assessment =
+      insert(:assessment, %{
+        open_at: Timex.shift(Timex.now(), days: 5),
+        close_at: Timex.shift(Timex.now(), days: 10)
+      })
+
+    before_open_at_question =
+      insert(:question, %{assessment: before_open_at_assessment, type: :multiple_choice})
+
+    before_open_at_conn = post(conn, build_url(before_open_at_question.id), %{answer: 5})
+    assert response(before_open_at_conn, 401) == "Assessment not open"
+    assert is_nil(get_answer_value(before_open_at_question, before_open_at_assessment, user))
+
+    after_close_at_assessment =
+      insert(:assessment, %{
+        open_at: Timex.shift(Timex.now(), days: -10),
+        close_at: Timex.shift(Timex.now(), days: -5)
+      })
+
+    after_close_at_question =
+      insert(:question, %{assessment: after_close_at_assessment, type: :multiple_choice})
+
+    after_close_at_conn = post(conn, build_url(after_close_at_question.id), %{answer: 5})
+    assert response(after_close_at_conn, 401) == "Assessment not open"
+    assert is_nil(get_answer_value(after_close_at_question, after_close_at_assessment, user))
+
+    unpublished_assessment = insert(:assessment, %{is_published: false})
+
+    unpublished_question =
+      insert(:question, %{assessment: unpublished_assessment, type: :multiple_choice})
+
+    unpublished_conn = post(conn, build_url(unpublished_question.id), %{answer: 5})
+    assert response(unpublished_conn, 401) == "Assessment not open"
+    assert is_nil(get_answer_value(unpublished_question, unpublished_assessment, user))
+  end
+
+  defp build_url(question_id) do
+    "/v1/assessments/question/#{question_id}/submit/"
+  end
+
+  defp get_answer_value(question, assessment, user) do
+    answer =
+      Answer
+      |> where(question_id: ^question.id)
+      |> join(:inner, [a], s in assoc(a, :submission))
+      |> where([a, s], s.student_id == ^user.id)
+      |> where([a, s], s.assessment_id == ^assessment.id)
+      |> Repo.one()
+
+    if answer do
+      case question.type do
+        :multiple_choice -> Map.get(answer.answer, "choice_id")
+        :programming -> Map.get(answer.answer, "code")
+      end
+    end
+  end
```

Comment:
You should test the cases when role is `:staff` and `:admin`

Reply:
added admin test. staff test is there but commented out because i currently allow staff to submit (my role is staff)

Reply:
Lmao. We should really make the mocking interface. I'll try to do it tonight, #104